### PR TITLE
[NVIDIA XLA:GPU] Introducing training support for cudnn fused mha (LHLO lowering and thunk, 2nd out of 3)

### PR DESCRIPTION
**Pull request by Tixxx**
This is a follow-up pr for [this](https://github.com/openxla/xla/pull/3886).
In total, fused MHA training changes are broken into 3 PRs.
This is the 2nd of 3. The other PRs' links:
1st - Stream executor changes are in this https://github.com/openxla/xla/pull/3886 (merged)
3rd - Rewriter changes are in this https://github.com/openxla/xla/pull/3726
This pr contains changes to lower lhlo fused mha custom call to cudnnFMHA thunk and runner for both forward and backward fmha calls. It also adds activation output to forward calls in thunk, runner and stream executor to support training cases.
All the functional and unit testing will be in the 3rd and final PR.

### CODE REVIEW

File: xla/stream_executor/dnn.cc [hunk 1]
```diff
@@ -200,6 +200,7 @@ DnnSupport::FusedMHASoftmaxRunnerFromDesc(
     const dnn::MatmulTensorDescriptor& bmm2_rhs_descriptor,
     const dnn::MatmulTensorDescriptor& intermediate_bmm2_lhs_descriptor,
     const dnn::TensorDescriptor& output_descriptor,
```
Comment (cheshire) [inline xla/stream_executor/dnn.cc:202]: Could SE changes be separated out?

**Review by cheshire — CHANGES_REQUESTED** [review]
Could we add tests?

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 2]
```diff
@@ -519,7 +519,31 @@ StatusOr<xla::gpu::CudnnfMHAKind> AsCudnnfMHAKind(
...
+    mlir::lmhlo_gpu::FusedMhaBackwardDagSignature signature) {
+  switch (signature) {
+    // backward
+    case mlir::lmhlo_gpu::FusedMhaBackwardDagSignature::
+        BackwardScaleBiasSoftmax:
+      return xla::gpu::CudnnfMHAKind::kBackwardScaleBiasSoftmax;
+    case mlir::lmhlo_gpu::FusedMhaBackwardDagSignature::
+        BackwardScaleBiasSoftmaxDropout:
+      return xla::gpu::CudnnfMHAKind::kBackwardScaleBiasSoftmaxDropout;
+      break;
```
Comment (cheshire) [inline xla/service/gpu/ir_emitter_unnested.cc:536]: Extraneous "break", here and below.

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 3]
```diff
@@ -393,6 +406,200 @@ void AssignSeed(GpufMHAConfig &config,
...
+    lazy_runner = &*local_runner;
+  }
+  // FMHA TODO: add GetDNNFusedMHAKindFromCudnnfMHAKind here
+  TF_ASSIGN_OR_RETURN(se::dnn::FusedMHAKind kind,
+                      GetDNNFusedMHAKindFromCudnnfMHAKind(params.config->kind));
+  std::optional<double> dropout_rate;
+  if (params.config->dropout_rate) dropout_rate = *params.config->dropout_rate;
+
+  double scale = 1.0;
+  if (params.config->fmha_scale) scale = *params.config->fmha_scale;
```
Comment (cheshire) [inline xla/service/gpu/gpu_fused_mha_runner.cc:438]: Code style prefers braces everywhere

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 4]
```diff
@@ -464,17 +683,207 @@ void AssignSeed(GpufMHAConfig &config,
...
+
+  auto assign_scale = [&]() {
+    config.fmha_scale.emplace();
+    double &fmha_scale = *config.fmha_scale;
+    fmha_scale = backend_config.fmha_scale();
+  };
+
+  auto assign_dropout_rate = [&]() {
+    config.dropout_rate.emplace();
+    double &dropout_rate = *config.dropout_rate;
```
[the remark this answers is not in the document]
Comment (cheshire) [reply xla/service/gpu/gpu_fused_mha_runner.cc:823]: I'm assuming config.dropout_rate = backend_config.dropout_rate() does not work?

File: xla/service/gpu/gpu_fused_mha_runner.cc (hunk 3 above)
Comment (Tixxx) [reply xla/service/gpu/gpu_fused_mha_runner.cc]: changed

**Review by cheshire — APPROVED** [review]

**Pull request closed without merging**